lib/distance/vincenty.ex

```diff
@@ -0,0 +1,191 @@
+defmodule Distance.Vincenty do
+  @moduledoc ~S"""
+  Calculate distance per [Vincenty's inverse formula](https://en.wikipedia.org/wiki/Vincenty%27s_formulae)
+  (shortest travel distance on the surface of an [oblate spheroid](https://en.wikipedia.org/wiki/Spheroid#Oblate_spheroids) Earth) given two longitude-latitude pairs.
+
+  This method is iterative and more costly than other methods, such as the [great circle](lib/distance/great_circle.ex) method, but also potentially more accurate.
+  It is important to note that [nearly antipodal points](https://en.wikipedia.org/wiki/Vincenty%27s_formulae#Nearly_antipodal_points) can cause convergence issues with this method.
+
+  The function accepts two tuples in the form of `{longitude, latitude}` and
+  returns the distance in meters. It will also accept a List of tuples.
+  """
+
+  @type coords() :: {number(), number()}
+
+  @convergence_threshold 1.0e-12
+  @ellipsoid_flattening 1 / 298.257223563
+  @max_iterations 200
+  @radians_in_degrees 180 / :math.pi()
+  @radius_at_equator 6_378_137
+  @radius_at_poles (1 - @ellipsoid_flattening) * @radius_at_equator
+
+  @doc """
+  Returns the distance in meters between two points in the form of
+  `{longitude, latitude}`, per Vincenty's inverse formula.
+
+  ## Examples
+      iex> Distance.Vincenty.distance({-105.343, 39.984}, {-105.534, 39.123})
+      96992.65430928342
+      iex> Distance.Vincenty.distance({-74.00597, 40.71427}, {-70.56656, -33.42628})
+      8216469.550643708
+  """
+
+  @spec distance(coords(), coords()) :: float()
+  def distance(coord, coord), do: 0.0
+
+  def distance({lon1, lat1}, {lon2, lat2}) do
+    u1 = calculate_reduced_latitude(lat1)
+    u2 = calculate_reduced_latitude(lat2)
+    lambda_initial = degrees_to_radians(lon2 - lon1)
+
+    sin_u1 = :math.sin(u1)
+    cos_u1 = :math.cos(u1)
+    sin_u2 = :math.sin(u2)
+    cos_u2 = :math.cos(u2)
+
+    {_, cos_squared_alpha, sin_sigma, cos_sigma, cos2_sigma_m, sigma} =
+      1..@max_iterations
+      |> Enum.reduce_while({lambda_initial, 0, 0, 0, 0, 0}, fn _, {lambda_prev, _, _, _, _, _} ->
+        sin_lambda = :math.sin(lambda_prev)
+        cos_lambda = :math.cos(lambda_prev)
+        sin_sigma = calculate_sin_sigma(sin_u1, sin_u2, sin_lambda, cos_u1, cos_u2, cos_lambda)
+        cos_sigma = calculate_cos_sigma(sin_u1, sin_u2, cos_u1, cos_u2, cos_lambda)
+        sigma = :math.atan2(sin_sigma, cos_sigma)
+        sin_alpha = calculate_sin_alpha(cos_u1, cos_u2, sin_lambda, sin_sigma)
+        cos_squared_alpha = calculate_cos_squared_alpha(sin_alpha)
+        cos2_sigma_m = calculate_cos2_sigma_m(sin_u1, sin_u2, cos_sigma, cos_squared_alpha)
+        c = calculate_c(cos_squared_alpha)
+
+        lambda =
+          calculate_lambda(
+            lambda_initial,
+            c,
+            sin_alpha,
+            sigma,
+            sin_sigma,
+            cos_sigma,
+            cos2_sigma_m
+          )
```

Comment:
I opted for several small helper functions to keep this more concise/readable, but the result is several helper functions that are only used once. I'm fairly ambivalent on this if it is preferred to just keep the logic in-line.

Reply:
I think this makes more sense than using a single large function.